processing/src/main/java/io/druid/segment/virtual/ExpressionObjectSelector.java

```diff
@@ -84,91 +88,61 @@ public static ExpressionObjectSelector from(ColumnSelectorFactory columnSelector
 
   @VisibleForTesting
   @Nonnull
-  static Supplier<Number> supplierFromFloatSelector(final FloatColumnSelector selector)
+  static Supplier<Object> supplierFromDimensionSelector(final DimensionSelector selector)
   {
     Preconditions.checkNotNull(selector, "selector");
-    return new Supplier<Number>()
-    {
-      @Override
-      public Number get()
-      {
-        return selector.get();
-      }
-    };
-  }
-
-  @VisibleForTesting
-  @Nonnull
-  static Supplier<Number> supplierFromLongSelector(final LongColumnSelector selector)
-  {
-    Preconditions.checkNotNull(selector, "selector");
-    return new Supplier<Number>()
-    {
-      @Override
-      public Number get()
-      {
-        return selector.get();
+    return () -> {
+      final IndexedInts row = selector.getRow();
+      if (row.size() == 0) {
+        // Treat empty multi-value rows as nulls.
+        return null;
+      } else if (row.size() == 1) {
+        return selector.lookupName(row.get(0));
+      } else {
+        // Can't handle multi-value rows in expressions.
+        // Treat them as nulls until we think of something better to do.
+        return null;
       }
     };
   }
 
   @VisibleForTesting
   @Nullable
-  static Supplier<Number> supplierFromObjectSelector(final ObjectColumnSelector selector)
+  static Supplier<Object> supplierFromObjectSelector(final ObjectColumnSelector selector)
   {
-    final Class<?> clazz = selector == null ? null : selector.classOfObject();
-    if (selector != null && (clazz.isAssignableFrom(Number.class)
-                             || clazz.isAssignableFrom(String.class)
-                             || Number.class.isAssignableFrom(clazz))) {
-      // There may be numbers here.
-      return new Supplier<Number>()
-      {
-        @Override
-        public Number get()
-        {
-          return tryParse(selector.get());
-        }
-      };
-    } else {
-      // We know there are no numbers here. Use a null supplier.
+    if (selector == null) {
       return null;
     }
-  }
 
-  @Nullable
-  private static Number tryParse(final Object value)
-  {
-    if (value == null) {
+    final Class<?> clazz = selector.classOfObject();
+    if (Number.class.isAssignableFrom(clazz) || String.class.isAssignableFrom(clazz)) {
+      // Number, String supported as-is.
+      return selector::get;
+    } else if (clazz.isAssignableFrom(Number.class) || clazz.isAssignableFrom(String.class)) {
+      // Might be Numbers and Strings. Use a selector that double-checks.
+      return () -> {
+        final Object val = selector.get();
+        if (val instanceof Number || val instanceof String) {
+          return val;
+        } else {
+          return null;
+        }
+      };
+    } else {
+      // No numbers or strings.
       return null;
```

Comment:
I guess it doesn't really matter, since a null supplier is treated equivalently to a supplier that always returns null.